agent/cache-types/streaming_health_services.go

```diff
@@ -22,170 +20,65 @@ const (
-	sub.lock.RLock()
-	if sub.lastResult.Index > opts.MinIndex {
-		result.Value = &sub.lastResult
-		result.Index = sub.lastResult.Index
-		sub.lock.RUnlock()
-		return result, nil
+	handler := healthServicesHandler{
+		state: make(map[string]structs.CheckServiceNode),
 	}
-	sub.lock.RUnlock()
-
-	// Wait for an update based on the index we're blocking on.
-	timeout := time.After(StreamTimeout)
-WAIT:
-	select {
-	case r := <-sub.resultCh:
-		switch reply := r.(type) {
-		case *structs.IndexedCheckServiceNodes:
-			if reply.Index <= opts.MinIndex {
-				goto WAIT
-			}
-			result.Value = reply
-			result.Index = reply.Index
-		case error:
-			result.Value = reply
+	indexFunc := func(v interface{}) uint64 {
+		if v, ok := v.(*structs.IndexedCheckServiceNodes); ok {
+			return v.Index
 		}
-	case <-timeout:
-		sub.lock.RLock()
-		result.Value = &sub.lastResult
-		result.Index = sub.lastResult.Index
-		sub.lock.RUnlock()
+		return 0
 	}
 
-	return result, nil
+	return watchSubscriber(c.Client, opts, subscribeReq, &handler, indexFunc)
 }
 
 func (c *StreamingHealthServices) SupportsBlocking() bool {
 	return true
 }
 
-// subscriber runs a subscription for a single topic/key combination.
-type subscriber struct {
-	client     stream.ConsulClient
-	service    string
-	lastResult structs.IndexedCheckServiceNodes
-	lastUpdate time.Time
-	resultCh   chan interface{}
-
-	cancelFunc func()
-
-	lock sync.RWMutex
-}
-
-// Close stops the subscriber by cancelling the context.
-func (s *subscriber) Close() error {
-	s.cancelFunc()
-	return nil
+// healthServicesHandler maintains a view of the health of a service
+// based on incoming events from a Subscriber.
+type healthServicesHandler struct {
+	state map[string]structs.CheckServiceNode
 }
 
-// run starts a Subscribe call for the subscriber's key, receives the initial
-// snapshot of state and then sends the state to the watcher as soon as its
-// watched index requirement is met.
-func (s *subscriber) run() {
-	ctx, cancel := context.WithCancel(context.Background())
-	s.cancelFunc = cancel
-	defer s.cancelFunc()
-
-	// Start a new Subscribe call.
-	streamHandle, err := s.client.Subscribe(ctx, &stream.SubscribeRequest{
-		Topic: stream.Topic_ServiceHealth,
-		Key:   s.service,
-	})
-
-	// If something went wrong setting up the stream, return the error.
-	if err != nil {
-		// Do a blocking send since there should always be a Fetch goroutine watching
-		// on the startup of a subscriber.
-		s.resultCh <- err
-		return
+// HandleEvent updates the handler's state based on register/deregister events.
+func (h *healthServicesHandler) HandleEvent(event *stream.Event) {
+	serviceHealth := event.GetServiceHealth()
+	node := serviceHealth.ServiceNode
+	id := fmt.Sprintf("%s/%s", node.Node.Node, node.Service.ID)
+
+	switch serviceHealth.Op {
+	case stream.CatalogOp_Register:
+		checkServiceNode := stream.FromCheckServiceNode(serviceHealth.ServiceNode)
+		h.state[id] = checkServiceNode
```

Comment:
This is being called from the Subscriber goroutine but State below is being accessed by the Fetch goroutine. I think we need to synchronize map accesses here.

Reply:
both are only called in Subscriber.run(), unless you're referring to something else?

Reply:
🤔 Think I was confused with the `State` variable in the Fetcher.